documentation/project-docs/cli-guidelines.md

```diff
@@ -0,0 +1,216 @@
+
+When this is used, you should load the sources list from the NuGet configuration (using the [libraries][nuget-configuration]) and create a
+ new `PackageSource` to add to that set. This should be additive and not destructive.
+
+#### Exclusive package sources
+
+Long form: `--source <source_uri>`
+
+When this is used, you should ONLY use the sources supplied by this parameter and ignore any sources in the NuGet configuration. This is useful
+for scenarios where you want to ensure that you're only using a specific source for a specific operation.
+
+### Feed Authentication support
+
+Long form: `--interactive <bool>`
+Default value: `true` when in an [interactive session](#determining-interactive-sessions), `false` otherwise
+
+NuGet authentication often requires some interactive action like going to a browser page. You should 'prime' the NuGet credential services by calling
+`NuGet.Credentials.DefaultCredentialServiceUtility.SetupDefaultCredentialService(ILogger logger, bool nonInteractive)` with the value of `nonInteractive`
+being the inverted value of this parameter. This will ensure that the credential service is set up correctly for the current session type.
+
+## Contextual behaviors
+
+### Implicit project/solution file discovery
+
+When a command is run in a directory that contains a project or solution file, the command should automatically use that file as the
+target for the command.
+If the command is run in a directory that contains multiple project or solution files, the command should fail with an error message
+indicating that the user should specify which file to use.
+
+Consider the use of an option like `--project <path to project or solution>` to allow for executing the command against a specific
+project or solution file at an arbitrary location - but be aware that many mechanisms in .NET are hierarchical from Current Working
+Directory and may have different behaviors when used in this detached mode.
+
+### Determining interactive sessions
+
+Commands should be able to determine if they are running in an interactive session or not. This is important for determining
+
+* whether to prompt the user for input or not,
+* what the default value for NuGet's `--interactive` option should be
+* how to format output (along with terminfo probing)
+
+In general the easiest way to check this is if the `stdin` file descriptor is a TTY, or in .NET Terms: `Console.IsInputRedirected` is `false`.
+
+## Interaction patterns
+
+### Support structured (JSON) output
+
+All commands should support structured (JSON) output. This is a key part of the CLI experience and allows for easier scripting and automation of the CLI.
+See the [`--format`](#output-modesformatting) option for more details. When structured output is requested, the command should output the structured
+data to `stdout` and any non-structured data to `stderr` - this is to make it easier to parse the structured data without having to deal with the
+non-structured data.
```

Comment:
I like this idea, but we don't have any common way to support this. It would be nice if the `--format` had common code to do the different format types. Many concepts like this are quite disparate in the CLI code.

Reply:
I'm less concerned with the technical side of how we'd do this - the doc is supposed to outline a set of expectactions/guidelines and then we separately work towards making it easy to adhere to them. I do agree this is currently difficult, and making some code-facing helpers or guidelines would help to reduce the pain.